File: core/src/com/github/servb/antonbird/gameworld/GameWorld.java
/*
 * Copyright 2018 SerVB.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.servb.antonbird.gameworld;

import com.badlogic.gdx.math.Intersector;
import com.badlogic.gdx.math.Rectangle;
import com.github.servb.antonbird.gameobject.Anton;
import com.github.servb.antonbird.gameobject.ScrollHandler;
import com.github.servb.antonbird.helper.AssetLoader;
import java.util.Random;

/**
 *
 * @author SerVB
 */
public final class GameWorld {

    private final Anton bird;
    private final ScrollHandler scroller;

    private final Rectangle ground;

    private int score = 0;

    public GameWorld(final int midPointY) {
        bird = new Anton(33, midPointY - 5, 17, 12);
        scroller = new ScrollHandler(this, midPointY + 66);
        ground = new Rectangle(0, midPointY + 66, 136, 11);
    }

    public final void update(float delta) {
        if (delta > .15f) {
            delta = .15f;
        }

        bird.update(delta);
        scroller.update(delta);

        if (scroller.collides(bird) && bird.isAlive()) {
            scroller.stop();
            bird.die();
            AssetLoader.dead[new Random().nextInt(AssetLoader.DIE_SOUND_COUNT)].play();
        }

        if (Intersector.overlaps(bird.getBoundingCircle(), ground)) {
            scroller.stop();
            bird.die();
            bird.decelerate();
        }
    }

    public final Anton getBird() {
        return bird;
    }

    public final ScrollHandler getScroller() {
        return scroller;
    }

    public int getScore() {
        return score;
    }

    public void addScore(final int increment) {
        score += increment;
    }
}

File: core/src/com/github/servb/antonbird/gameworld/GameWorld.kt
package com.github.servb.antonbird.gameworld

import com.badlogic.gdx.math.Intersector
import com.badlogic.gdx.math.Rectangle
import com.github.servb.antonbird.gameobject.Anton
import com.github.servb.antonbird.gameobject.ScrollHandler
import com.github.servb.antonbird.helper.AssetLoader

class GameWorld(private val midPointY: Int) {

    val bird: Anton = Anton(33f, midPointY - 5f, 17, 12)
    val scroller: ScrollHandler = ScrollHandler(this, midPointY + 66f)

    private val ground: Rectangle = Rectangle(0f, midPointY + 66f, 136f, 11f)

    var score = 0
        private set

    private var currentState = GameState.READY

    val isReady: Boolean get() = currentState == GameState.READY
    val isGameOver: Boolean get() = currentState == GameState.GAME_OVER
    val isHighScore: Boolean get() = currentState == GameState.HIGH_SCORE

    fun start() {
        currentState = GameState.RUNNING
    }

    fun restart() {
        currentState = GameState.READY
        score = 0
        bird.onRestart(midPointY - 5f)
        scroller.onRestart()
    }

    fun update(delta: Float) = when (currentState) {
        GameState.READY -> updateReady(delta)
        GameState.RUNNING -> updateRunning(delta)
        else -> Unit
    }

    private fun updateReady(delta: Float) {
    }

    private fun updateRunning(delta: Float) {
        @Suppress("NAME_SHADOWING") val delta = minOf(delta, 0.15f)

        bird.update(delta)
        scroller.update(delta)

        if (scroller.collides(bird) && bird.isAlive) {
            scroller.stop()
            bird.die()
            AssetLoader.dead.random().play()
        }

        if (Intersector.overlaps(bird.boundingCircle, ground)) {
            scroller.stop()
            bird.die()
            bird.decelerate()

            if (score > AssetLoader.highScore) {
                AssetLoader.highScore = score
                currentState = GameState.HIGH_SCORE
            } else {
                currentState = GameState.GAME_OVER
            }
        }
    }

    fun addScore(increment: Int) {
        score += increment
    }

    enum class GameState {
        READY,
        RUNNING,
        GAME_OVER,
        HIGH_SCORE,
    }
}
